util/exercise/src/structs.rs

```diff
@@ -0,0 +1,159 @@
+use serde::{Deserialize, Serialize};
+
+macro_rules! pub_struct_serde_getters {
+    (
+        pub struct $name:ident {
+            $( $fname:ident : $ftype:ty ),* $(,)*
+        }
+    ) => {
+        #[derive(Serialize, Deserialize, Debug)]
+        pub struct $name {
+            $($fname : $ftype),*
+        }
+
+        impl $name {
+            $(
+            pub fn $fname(&self) -> &$ftype {
+                &self.$fname
+            }
+            )*
+        }
+    };
+}
+
+pub_struct_serde_getters! {
+    pub struct CanonicalData {
+        exercise: Exercise,
+        version: Version,
+        comments: Option<Comments>,
+        cases: TestGroup,
+    }
+}
+
+type Exercise = String;
+type Version = String;
+type Comments = Vec<String>;
+type TestGroup = Vec<LabeledTestItem>;
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum LabeledTestItem {
+    Single(LabeledTest),
+    Array(LabeledTestGroup),
+}
+
+pub_struct_serde_getters! {
+    pub struct LabeledTest {
+        description: Description,
+        optional: Option<Optional>,
+        comments: Option<Comments>,
+        property: Property,
+        input: Input,
+        expected: Expected,
+    }
+}
+
+impl LabeledTest {
+    pub fn generate_test_function(&self) -> String {
+        // Attributes.
+        let mut s = vec![String::from("#[test]"), String::from("#[ignore]")];
+
+        // Description and comments.
+        s.push(format!("/// {}", self.description()));
+        if let Some(comments) = self.comments() {
+            for line in comments {
+                s.push(format!("/// {}", line));
+            }
+        }
+
+        // Function signature (plus opening bracket).
+        s.push(format!(
+            "fn test_{}() {{",
+            exercise::format_exercise_description(self.description())
+        ));
+
+        // Unpack the inputs.
+        let mut inputs = Vec::new();
+        for (name, value) in self.input() {
+            inputs.push(name.clone());
+            s.push(format!("let {} = {};", name, value.to_literal()));
+        }
+        let inputs = inputs.join(", ");
+
+        // Output.
+        s.push(format!("let expected = {};", self.expected.to_literal()));
+
+        // Add call to the processing function (and closing brackets).
+        s.push(format!(
+            "process_{}_case({}, expected);\n}}\n",
+            exercise::format_exercise_property(self.property()),
+            inputs
+        ));
+
+        // Add newlines to pretty it up.
+        s.join("\n")
+    }
+}
+
+pub_struct_serde_getters! {
+    pub struct LabeledTestGroup {
+        description: Description,
+        optional: Option<Optional>,
+        comments: Option<Comments>,
+        cases: TestGroup,
+    }
+}
+
+type Description = String;
+type Optional = String;
+type Property = String;
+type Input = IndexMap<String, Data>;
+type Expected = Data;
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum Data {
```

Comment:
It started with a lot of good intentions. :laughing: 

It is interesting that we can deserialize into `i32`, `f64`, `char`, `String`, etc. The idea (hope) was that the `Map` variant would be a `IndexMap<Data, Data>`, which would give the type information; for example, for `alphametics`, the expected value as a `IndexMap<char, i32>` (and possibly `u8` eventually). It kinda works, but on a case-by-case basis (no pun intended); so we end up with a `IndexMap<char, i32>` for one test case and a `IndexMap<String, i32>` for another under the same property. Alas, my hopes were dashed. :sob: 

This is not to say that it's impossible, but I don't think asking `serde` to deliver this directly is possible. That being said, and thinking out loud, one possible way I've been mulling over is testing repeatedly `serde_json::value::Value`'s `is_` methods against all the inputs, and taking the "last" (or largest, etc.) type that fit all of them. As you mention, a newtype might help by implementing `PartialOrd` over the variants. But it does feel a bit hacky.

I do like the idea of a sub-enum, but I fear that it might encounter the issue related to case-by-case deserialization. That being said, it would probably be easier to implement the heuristic to find the best fit.

Reply:
I've been thinking about this. Since `inputs` are always JSON objects, which we translate into `IndexMap`s, comparing keys across all the cases for a given property, using `HashSet<String>::symmetric_difference` for example, would tell us if we can use a struct or not. 

If they all share the same keys, and if we assume that for a the values for a given key all share the same type, unpacking into a struct is relatively straightforward. The assumption that the values are of the same type seems reasonable, but I have been burned in the past.

If they don't share the same keys, then a struct would require using `Option`s, and, at this point, a `HashMap` is probably the simpler choice. For example, in [`rest_api`](https://github.com/exercism/problem-specifications/blob/master/exercises/rest-api/canonical-data.json), the first case has no payload while the third one (the second one using `post`) does.

If we want to do this recursively, I think that using our own enum is probably the simplest solution (though maybe with a better name than `Data` :wink: ). It's not necessary an easy task, but I think it essentially boils down to using level order traversal.

Reply:
> If they all share the same keys, and if we assume that for a the values for a given key all share the same type, unpacking into a struct is relatively straightforward.

Agree, that seems a reasonable heuristic.

> If they don't share the same keys, [...] a HashMap is probably the simpler choice. 

Agree.

> If we want to do this recursively, ...

I'm not opposed to an iterative approach if one feels natural; it just seemed to me that a recursive traversal would be most natural, because we need to do type-detection and analysis on arbitrary sub-types. If (making this up) an exercise has `"input": {"a": {"b": "c"}}`, then we need to check that (bare minimum) all `input.a` also heuristically identify as structs (i.e. they all have a `b` whose value type is stringy) in order to categorize the whole `input` type as a struct instead of a map.

That said, any kind of type detection is going to end up being a vast improvement over the current state of affairs, which unconditionally uses maps, even when they're really not appropriate.

Reply:
> I'm not opposed to an iterative approach if one feels natural; it just seemed to me that a recursive traversal would be most natural, because we need to do type-detection and analysis on arbitrary sub-types.

Agreed.

> If (making this up) an exercise has "input": {"a": {"b": "c"}}, then we need to check that (bare minimum) all input.a also heuristically identify as structs (i.e. they all have a b whose value type is stringy) in order to categorize the whole input type as a struct instead of a map.

I think I agree, but I'm not 100 % sure. Let's say `input.a` can't be reduced to a struct (sometimes the only key is `b`, sometimes `d`), so it has to "remain" a map. Couldn't we unpack it into a struct, `(Map<_, _>, String, String)` for example?